scripts/nn/layers/conv2d_transpose.dml

```diff
@@ -0,0 +1,172 @@
+   *  - b: Biases, of shape (F, 1).
+   *  - C: Number of input channels (dimensionality of depth).
+   *  - Hin: Input height.
+   *  - Win: Input width.
+   *  - Hf: Filter height.
+   *  - Wf: Filter width.
+   *  - strideh: Stride over height.
+   *  - stridew: Stride over width.
+   *  - padh: Padding for top and bottom sides.
+   *  - padw: Padding for left and right sides.
+   *
+   * Outputs:
+   *  - dX: Gradient wrt `X`, of shape (N, C*Hin*Win).
+   *  - dW: Gradient wrt `W`, of shape (F, C*Hf*Wf).
+   *  - db: Gradient wrt `b`, of shape (F, 1).
+   */
+  N = nrow(X)
+  F = nrow(W)
+
+  dW = conv2d_backward_filter(dout, X, stride=[strideh,stridew], padding=[padh,padw],
+                              input_shape=[N,C,Hout,Wout], filter_shape=[F,C,Hf,Wf])
+
+  dX = conv2d(dout, W, input_shape=[N,C,Hout,Wout], filter_shape=[F,C,Hf,Wf],
+              stride=[strideh,stridew], padding=[padh,padw])
+
+  db = rowSums(matrix(colSums(dout), rows=F, cols=Hout*Wout))
+}
+
+init = function(int F, int C, int Hf, int Wf)
+    return (matrix[double] W, matrix[double] b){
+  /*
+   * Utility function to initialize the parameters of this layer.
+   *
+   * We use the heuristic by He et al., which limits the magnification
+   * of inputs/gradients during forward/backward passes by scaling
+   * unit-Gaussian weights by a factor of sqrt(2/n), under the
+   * assumption of relu neurons.
+   *  - http://arxiv.org/abs/1502.01852
+   *
+   * Inputs:
+   *  - F: Number of filters.
+   *  - C: Number of input channels (dimensionality of depth).
+   *  - Hf: Filter height.
+   *  - Wf: Filter width.
+   *
+   * Outputs:
+   *  - W: Weights, of shape (F, C*Hf*Wf).
+   *  - b: Biases, of shape (F, 1).
+   */
+  W = rand(rows=F, cols=C*Hf*Wf, pdf="normal") * sqrt(2/(C*Hf*Wf))
+  b = matrix(0, rows=F, cols=1)
+}
+
+init_bilinear = function(int C, int K)
+    return (matrix[double] W, matrix[double] b){
+  /*
+   * Utility function to upsample using this layer.
+   *
+   * Upsampling the input by factor f (each side) requires
+   * channel-wise independent kernels of size K = 2f - f%2,
+   * stride = f and pad = ceil((f-1)/2). The weights are set
+   * via bilinear interpolation, bias is set to 0.
+   * 
+   * Inputs:
+   *  - C: Number of input channels (dimensionality of depth).
+   *  - K: Kernel size (upsampling requires a square filter
+   *      of size K X K).
+   *
+   * Outputs:
+   *  - W: Weights, of shape (F, C*Hf*Wf).
+   *  - b: Biases, of shape (F, 1).
+   */
+  factor_up = ceil(K / 2)
+  center = (2 * factor_up - factor_up %% 2 - 1) / 2 / factor_up
+  vect = 1 - abs(seq(0, K-1) / factor_up - center)
+  weights = matrix(vect %*% t(vect), rows=1, cols=K*K)
+
+  if(C > 1){
+    repl_weights = matrix(1, rows=C, cols=1) %*% cbind(weights, matrix(0, rows=1, cols=C*K*K))
+    repl_weights_row = matrix(repl_weights, rows=1, cols=C*(C+1)*K^2)
+    W = matrix(repl_weights_row[1,1:(C*K)^2], rows=C, cols=C*K^2)
+  }else W = weights
```

Comment:
Actually, I take that back. The 'if' condition can be commented out. The lines within the 'if' block will work even in the case of C=1. It just results in redundant computation (I don't think the optimizer is smart enough to recognize this). On the other hand, I don't think this bit of redundant computation will ever result in a bottleneck.